File: app/src/main/java/com/sq26/experience/service/AIDLService.java
package com.sq26.experience.service;

import android.app.Service;
import android.content.Intent;
import android.os.IBinder;
import android.os.RemoteException;

import androidx.annotation.Nullable;

import com.sq26.experience.aidl.IAidlInterface;
import com.sq26.experience.aidl.ICallbackAidlInterface;

import java.util.Map;
import java.util.Objects;
import java.util.concurrent.ConcurrentHashMap;

public class AIDLService extends Service {

    private Map<String, ICallbackAidlInterface> map = new ConcurrentHashMap<>();

    private IAidlInterface.Stub iAidlInterface = new IAidlInterface.Stub() {
        @Override
        public void pushText(String text) throws RemoteException {
            for (String key : map.keySet()) {
                Objects.requireNonNull(map.get(key)).pull(text);
            }
        }

        @Override
        public void registerCallback(String packageName, ICallbackAidlInterface callback) throws RemoteException {
            map.put(packageName, callback);
        }

        @Override
        public void unregisterCallback(String packageName) throws RemoteException {
            map.remove(packageName);
        }
    };


    @Nullable
    @Override
    public IBinder onBind(Intent intent) {

        return iAidlInterface;
    }
}

File: app/src/main/java/com/sq26/experience/service/AIDLService.kt
package com.sq26.experience.service

import android.app.Service
import android.content.Intent
import android.os.IBinder
import android.os.RemoteException
import com.sq26.experience.aidl.IAidlInterface
import com.sq26.experience.aidl.ICallbackAidlInterface
import java.util.concurrent.ConcurrentHashMap

class AIDLService : Service() {
    //定义一个线程安全的map用于保存通讯回调
    private val map: MutableMap<String, ICallbackAidlInterface> = ConcurrentHashMap()

    //实现aidi接口
    private val iAidlInterface: IAidlInterface.Stub = object : IAidlInterface.Stub() {
        @Throws(RemoteException::class)
        override fun pushText(text: String) {
            //发送信息
            for (key in map.keys) {
                map[key]?.pull(text)
            }
        }

        @Throws(RemoteException::class)
        override fun registerCallback(name: String, callback: ICallbackAidlInterface) {
            //注册信息接收回调
            map[name] = callback
        }

        @Throws(RemoteException::class)
        override fun unregisterCallback(name: String) {
            //注销信息接收回调
            map.remove(name)
        }
    }

    override fun onBind(intent: Intent): IBinder {
        //返回服务接口实现
        return iAidlInterface
    }
}